sql/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/raster.scala

```diff
@@ -0,0 +1,223 @@
+    val factory = new GeometryFactory
+    val polygon = JTS.transform(factory.createPolygon(polyCoordinates), targetCRS)
+
+    polygon
+
+  }
+}
+
+
+// Fetches polygonal coordinates from a raster image
+case class ST_GeomFromRaster(inputExpressions: Seq[Expression])
+  extends Expression with CodegenFallback with UserDataGeneratator {
+  override def nullable: Boolean = false
+
+  override def eval(inputRow: InternalRow): Any = {
+    // This is an expression which takes one input expressions
+    assert(inputExpressions.length == 1)
+    val geomString = inputExpressions(0).eval(inputRow).asInstanceOf[UTF8String].toString
+    val geomConstruction = new GeometryOperations
+    val geometry = geomConstruction.readGeometry(geomString)
+    new GenericArrayData(GeometrySerializer.serialize(geometry))
+  }
+
+
+  override def dataType: DataType = GeometryUDT
+
+  override def children: Seq[Expression] = inputExpressions
+}
+
+
+// Constructs a raster dataframe from a raster image which contains multiple columns such as Geometry, Band values etc
+case class ST_GeomWithBandsFromGeoTiff(inputExpressions: Seq[Expression])
+  extends Expression with CodegenFallback with UserDataGeneratator {
+  override def nullable: Boolean = false
+
+  override def eval(inputRow: InternalRow): Any = {
+    // This is an expression which takes one input expressions
+    assert(inputExpressions.length == 2)
+    val geomString = inputExpressions(0).eval(inputRow).asInstanceOf[UTF8String].toString
+    val totalBands = inputExpressions(1).eval(inputRow).asInstanceOf[Int]
+    val geomConstruction = new GeometryOperations
+    val geometry = geomConstruction.readGeometry(geomString)
+    val bandvalues = getBands(geomString, totalBands)
+    returnValue(geometry.toGenericArrayData,bandvalues, 2)
+  }
+
+  private  def getBands(url: String, bands:Int): Array[Double] = {
+    val policy: ParameterValue[OverviewPolicy] = AbstractGridFormat.OVERVIEW_POLICY.createValue
+    policy.setValue(OverviewPolicy.IGNORE)
+
+    val gridsize: ParameterValue[String] = AbstractGridFormat.SUGGESTED_TILE_SIZE.createValue
+
+    val useJaiRead: ParameterValue[Boolean] = AbstractGridFormat.USE_JAI_IMAGEREAD.createValue.asInstanceOf[ParameterValue[Boolean]]
+    useJaiRead.setValue(true)
+
+
+    val reader: GridCoverage2DReader = new GeoTiffReader(url)
+    val coverage: GridCoverage2D = reader.read(Array[GeneralParameterValue](policy, gridsize, useJaiRead))
+
+    val dimensions: GridEnvelope = reader.getOriginalGridRange
+    val maxDimensions: GridCoordinates = dimensions.getHigh
+    val w: Int = maxDimensions.getCoordinateValue(0) + 1
+    val h: Int = maxDimensions.getCoordinateValue(1) + 1
+    val numBands: Int = bands
+
+    val bandValues: util.List[util.List[Double]] = new util.ArrayList[util.List[Double]](numBands)
+
+    for (i <- 0 until numBands) {
+      bandValues.add(new util.ArrayList[Double])
+    }
+
+    for (i <- 0 until w) {
```

Comment:
I believe this loop and the next loop can be merged

Reply:
@jiayuasu  The current loop returns an array of size equal to the number of bands for every pixel and I use it to construct a 2D array of size number of bands * number of pixels while the second loop flattens this array. If I need to merge both the loops, I have to keep a separate starting index for every band which may vary.